### DISCUSSION TITLE
We need a new image viewer

### DISCUSSION THREAD

**Discussion by benjaminhwilliams** [dials/dials#1428] [Ideas]
I have recently delved into the code for the `dials.image_viewer`, resulting in #1385, #1387 & #1388 and have discovered there are more than a few skeletons in the closet.

In an era when there are plenty of developer-friendly solutions to plotting image data interactively (Matplotlib widgets spring to mind as one possible solution, though there are many others that look slicker), I think it's worth taking a fresh look at whether wxPython provides the functionality we require at an acceptable development cost and also whether our use of a copied-and-pasted outdated version of PySlip (a package designed specifically for cartography and therefore perhaps not best adapted to our use case) is something we can afford to support.  As a newer member of this collaboration, I'm finding it difficult to decipher the code or to find documentation of decisions made in the current design of the image viewer.

My particular focus has been small-molecule crystallography data from synchrotron sources.  Having followed the trail that @Anthchirp pointed me to in #66, I can see that the image viewer is fundamentally hobbled for drawing ice rings, or any kind of isoresolution line, if the detector normal is not close to the beam direction.  This limitation seems to stem from design choices made when DIALS was young and the anticipated diffractometer geometries were simple.  If we want a fully functional image viewer for the case of non-zero detector two-theta (and since I'm responsible for supporting chemical crystallographers at Diamond, I would vote that we do), then I think we need to rebuild the image viewer from the ground up.

Above all, I'd like to avoid cases of the image viewer being allowed to accrue unnoticed bugs for many releases before anyone deals with them.  We shouldn't be subjecting our users to a tool that is not systematically tested.  We need an image viewer built from components that have unit tests, and which has a structure intelligible to a lay developer, otherwise we risk wasting our time supporting ossified code instead of supporting science.

**Comment by graeme-winter** [comment]
I think there has been a growing consensus (at least, amongst those I speak with) that the image viewer is in dire need of a complete overhaul so I completely agree with this. 

I would further suggest that this is quite possibly a solved problem in the outside world i.e. if you have arrays of data which you would like to slice & dice, overlay etc. and would like to look at. Modern Python, as alluded to above, contains much that is very relevant here. 

The current image viewer is completely broken now for our use cases at Diamond. Coming up with a new one however is likely to be expensive.

**Comment by phyy-nx** [comment]
I remember in 2013 seeing an image viewer slapped together by @hattne in OpenGL that showed the CSPAD in 3D, including any panel tilts, and it was a zippy little thing.  That's what I think we need.  Something that renders textures in arbitrary orientations, understands 3D objects, vectors, and curves from the ground up, uses an open 3D engine cross-compatible on the major operating systems and available in conda, and renders an orthographic projection by default.

**Comment by rjgildea** [comment]
I'm not convinced that user need/want a 3D image viewer. For example, it would certainly look impressive in demonstrations showing off the curved DLS I23 detector in a 3D image viewer, however it would be useless for actually looking at diffraction images, which is what users want in an image viewer. We already dials.geometry_viewer for examining the experimental geometry in 3D. I'm not sure how viewing diffraftion images in 3D aids their interpretation.

**Comment by phyy-nx** [comment]
It's not that the 3D viewer aids in interpretation, but what it does do is guarantee an accurate orthographic projection.  No off-by-one-half-pixel style coordinate conversion calls are needed since the viewer is already a 3D device and lab coordinates can be used directly.  The user would only ever see 2D (unless we provide a 3D checkmark), but the 3D engine ensures accuracy and speed.

**Comment by rjgildea** [comment]
What would an othrographic projection of a curved detector like the DLS I23 look like? What the user wants to see is the unrolled modules on the detector surface. I don't understand how an orthographic projection would achieve this.

**Comment by phyy-nx** [comment]
What does it look like now in the viewer?  I actually don't know :)  Regardless I can imagine several modes being useful that could be toggled between.  Perhaps the format class itself could provide hints as to the best way to view the data.  The point is, if the underlying toolkit is 3D, then math for drawing ice rings and such won't have to go through extra layers of coordinate transforms.  At least, that's how I see it.

**Comment by rjgildea** [comment]
You can actually see for yourself:
`$ dials.image_viewer $(libtbx.find_in_repositories dials_regression)/image_examples/DLS_I23/germ_13KeV_0001.cbf`
<img width="338" alt="Screenshot 2020-09-24 at 22 54 55" src="https://user-images.githubusercontent.com/2810624/94204384-0edd2400-feb9-11ea-84b8-2dfb4670f487.png">

This detector is actually handled as a special case in the image viewer code (not exactly ideal):
https://github.com/dials/dials/blob/414c84f0b7cde2effcc50b3af2ed813da9c6c05d/util/image_viewer/slip_viewer/tile_generation.py#L133-L155

If I disable that block of code with:
```
$ git diff
diff --git a/util/image_viewer/slip_viewer/tile_generation.py b/util/image_viewer/slip_viewer/tile_generation.py
index d57be26b1..4129df0b8 100644
--- a/util/image_viewer/slip_viewer/tile_generation.py
+++ b/util/image_viewer/slip_viewer/tile_generation.py
@@ -129,7 +129,7 @@ def get_flex_image_multipanel(
             panel.get_pixel_size()[1] * 1e-3,
         )
 
-        if len(panels) == 24 and panels[0].get_image_size() == (2463, 195):
+        if 0 and len(panels) == 24 and panels[0].get_image_size() == (2463, 195):
             rawdata.matrix_paste_block_in_place(
                 block=data.as_double(), i_row=i * data_padded[0], i_column=0
             )
```
then the panels are for some reason so far apart that only one of them will fit on screen at once:
<img width="663" alt="Screenshot 2020-09-24 at 23 00 28" src="https://user-images.githubusercontent.com/2810624/94204797-d9850600-feb9-11ea-84d6-ca7e3608b774.png">

**Comment by phyy-nx** [comment]
Thanks!  I was wondering if there was a special case.

**Comment by graeme-winter** [comment]
TL;DR: image viewers are for looking at images with analysis models superimposed, think of the (remote) user

# Long version

I'm coming from a different viewpoint here - I _like_ my image viewers to show images, with overlays, in the image coordinate system - my images are my data, and everything else should be adapted to model the observations. That said, I also recognise the potential value in a 3D inspection, but I am old and cranky and I would use the numbers for e.g. panel orientation to get that information. 

Now, there is another factor to chuck into the mix. 

Users. 

Users using facilities. 

I think we can no longer assume they are parked in front of a high performance beamline workstation a dozen feet from the detector. They're in far away places with strange sounding names, connecting via NX or whatever, quite possibly from their home broadband link, from a MacBook pro (on a good day) - I don't think we can assume people want or need (or can afford the bandwidth for) a whizzy 3D basis for their 2D image viewer. I know there _could_ be a vaccine before the year is out, but I doubt it and more to the point I actually don't see us widely going back to in-person experiments. 

Performance is also something people have complained about a _lot_. Simply blitting image data to an array for display with vector overlays should be quick, especially if we already encode up front how the 2D presentation of the image should look (https://github.com/cctbx/dxtbx/issues/17 👀)

So I recognise the request for something which shows the images in some kind of laboratory frame, but I don't think this can be the primary development target, at least in my humble opinion. Folding the images into the geometry viewer as @rjgildea mentions above is possibly a better solution for that? 

Finally, if you really want to show the "true" experiment you would need to put the viewpoint at the experiment origin and transform everything such that you are looking along the "diffracted rays" - since that is our model, that is what we are refining our understanding against. I don't think it would give a great deal of insight but it would look awesome.

**Comment by dagewa** [comment]
Note I think it would be trivially easy in most cases to obtain a 2D representation of panels for image viewer display
https://github.com/dials/dials/issues/1358#issuecomment-664853641
Non-coplanar cases like I23 need only override one method.

**Comment by phyy-nx** [comment]
Hi folks, I propose Thursday Nov 19th, 8am PST as a Zoom meeting to discuss the image viewer and a possible replacement for it.  Thumbs up on this post if you can make it.  If I get 5 upvotes I'll schedule it!

**Comment by nksauter** [comment]
Just to be clear I cannot endorse using Berkeley funds to collaborate on
this project.
Nick

Nicholas K. Sauter, Ph. D.
Senior Scientist, Molecular Biophysics & Integrated Bioimaging Division
Lawrence Berkeley National Laboratory
1 Cyclotron Rd., Bldg. 33R0345
Berkeley, CA 94720
(510) 486-5713

**Comment by phyy-nx** [comment]
To clarify, this isn't a funding discussion or any commitment of resources.  What we want to do in this meeting is simply capture a list of features that we would want in an image viewer, and leave discussion as to who does the work using which funding for another time.

**Comment by nksauter** [comment]
Yep, that  sounds like a good plan!

Nicholas K. Sauter, Ph. D.
Senior Scientist, Molecular Biophysics & Integrated Bioimaging Division
Lawrence Berkeley National Laboratory
1 Cyclotron Rd., Bldg. 33R0345
Berkeley, CA 94720
(510) 486-5713

**Comment by graeme-winter** [comment]
The brief discussion we had today concluded that we need to do something here, but no one has the necessary effort as it’s a major undertaking. Aim of meeting to try and find a way forward.

**Comment by phyy-nx** [comment]
Calendar invite sent to dials-support email list.  Ping me if you want it forwarded to you.

**Comment by phyy-nx** [comment]
Notes from zoom meeting today:

- List of improvements to current image viewer
    - Review and merge #1463
    - Checkbox: view in image frame vs laboratory frame.
    - .dials/imageview.conf → persistent settings on a per-user basis
    - show_diff_phil:  make it false? Figure out why it’s called twice? This is a performance increase
    - load_models = False should be the default
- Future image viewer
    - Core workflow
        - Fast zoom, pan, and image transition
        - Fast overlays of metadata
        - Fast addition of mathematical transforms of data (brightness/contrast, spotfinding thresholds, raw/corrected, sum/mean/max, difference images)
    - Three layers
        - Dxtbx/dials: knows how to read metadata into experiment lists, imagesets with get raw/corrected data, and reflection tables
        - Intermediate layer: takes above input and gives back bitmaps and vectors.  Should output ‘something standard’ that extant rendering engines can consume.
        - Rendering engine.  Purely computer science problem
    - This approach separates rendering engine from delivery of data.  Lets us have different front ends, like web or desktop applications.
    - Threaded read-ahead and zoom (zoom might be free from the rendering engine)
